src/test/java/uk/sky/cirrus/locking/LockVerificationTest.java

```diff
@@ -0,0 +1,209 @@
+
+    private static final int CASSANDRA_PORT = 9042;
+    private static final String KEYSPACE = "locker";
+    private static String TABLE_NAME = KEYSPACE + "." + "lock_testing";
+
+    private Cluster cluster;
+    private Session session;
+
+    @Before
+    public void setUp() {
+        cluster = createCluster();
+        session = cluster.connect();
+
+        session.execute(String.format("DROP KEYSPACE IF EXISTS %s;", KEYSPACE));
+        session.execute(String.format("DROP KEYSPACE IF EXISTS %s;", "locks"));
+        session.execute(String.format("CREATE KEYSPACE %s WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 3};", KEYSPACE));
+        Uninterruptibles.sleepUninterruptibly(300, TimeUnit.MILLISECONDS);
+        session.execute(String.format("CREATE TABLE %s (id text PRIMARY KEY, counter int);", TABLE_NAME));
+        Uninterruptibles.sleepUninterruptibly(800, TimeUnit.MILLISECONDS);
+        session.execute(new SimpleStatement(String.format("INSERT INTO %s (id, counter) VALUES (?, ?);", TABLE_NAME), "lock-tester", 1).setConsistencyLevel(ConsistencyLevel.QUORUM));
+    }
+
+    @After
+    public void cleanUp() {
+        final String dropKeyspaceQuery = String.format("DROP KEYSPACE IF EXISTS %s;", KEYSPACE);
+        session.execute(dropKeyspaceQuery);
+        session.close();
+        cluster.close();
+    }
+
+    @Test
+    public void shouldManageContentionsForSchemaMigrate() throws InterruptedException, URISyntaxException {
+
+        final Path cql_bootstrap = Paths.get(ClassLoader.getSystemResource("cql_migrate_multithreads").toURI());
+        final CqlMigratorImpl cqlMigrator = new CqlMigratorImpl(LockConfig.builder().build());
+
+        final Collection<Path> cqlPaths = singletonList(cql_bootstrap);
+        final ExecutorService cqlMigratorManager = newFixedThreadPool(25);
+        final Callable<String> cqlMigrate = () ->  { cqlMigrator.migrate(session, "cqlmigrate_test", cqlPaths); return "Done"; };
+        final List<Callable<String>> workers = newArrayList();
+
+        session.execute("DROP KEYSPACE IF EXISTS cqlmigrate_test;");
+
+        for(int i=0; i< 20; i++){
+            workers.add(cqlMigrate);
+        }
+
+        Stopwatch executionTime = Stopwatch.createStarted();
+        final List<Future<String>> futures = cqlMigratorManager.invokeAll(workers);
+
+        System.out.println(futures.size());
+
+        futures.forEach(future -> {
+            try {
+                future.get();
+            } catch (Exception e) {
+                e.printStackTrace();
+                Assert.fail(e.getMessage());
+            }
+        });
+
+        executionTime.stop();
+
+        cqlMigratorManager.shutdown();
+
+        final long elapsed = executionTime.elapsed(TimeUnit.SECONDS);
+        assertThat("Schema migration took longer than 25s", elapsed, Matchers.lessThanOrEqualTo(25L));
+    }
+
+    @Test
+    public void shouldObtainLockToInsertRecord() throws InterruptedException {
```

Comment:
This test was quite difficult for me to follow and also raised some concerns.    We are defining a callable that can returns state, but then we are also storing state in a shared global resource...  It also seems overly complicated for what we wanted to achieve.  

A much simpler approach would involve no coordination or shared state between running threads and would look     roughly as follows:

1. Define a Callable that 
1.1. increments a counter in the DB until it has reached X
1.2. returns a list of all the numbers it incremented
2. Block until all threads have completed their execution 
3. Verify Results
3.1.  The aggregated number of increments = X ( e.g. add all results to the list and size == X)
3.2.  No two threads incremented the same value. (e.g.  iterate from 0-X checking values match)

This approach be a lot less risky and involves no coordination or shared state.